osidb/api_views.py

```diff
@@ -96,6 +97,123 @@ def perform_update(self, serializer) -> None:
         self.perform_save(serializer)
 
 
+class ValidateModelViewSetMixin(ModelViewSetMixin):
+    # mixin class defining mechanism of API validations
+    # which are unlike model validations tied to query itself
+    #
+    # TODO this is not a proper class doc string as it would generate
+    # a ton of unrelated API endpoint descriptions in OpenAPI schema
+
+    _mandatory_attrs = None
+    _mandatory_params = None
+
+    def _validate_mandatory_attrs(self):
+        """
+        validate that all mandatory data attributes were provided
+        """
+        if self._mandatory_attrs is None:
+            return
+
+        for attr in self._mandatory_attrs:
+            if self.request.data.get(attr) is None:
+                raise BadRequest(f"Mandatory data attribute missing: {attr}")
+
+    def _validate_mandatory_params(self):
+        """
+        validate that all mandatory query params were provided
+        """
+        if self._mandatory_params is None:
+            return
+
+        for param in self._mandatory_params:
+            if self.request.query_params.get(param) is None:
+                raise BadRequest(f"Mandatory query parameter missing: {param}")
+
+    def validate(self) -> None:
+        """
+        validate query
+        raises BadRequest
+        """
+        for validation_name in [
+            item for item in dir(self) if item.startswith("_validate_")
+        ]:
+            # run every defined validation
+            getattr(self, validation_name)()
+
+    def create(self, request, *args, **kwargs) -> Response:
+        """
+        create model instance if valid
+        """
+        self.validate()
+        return super().create(request, *args, **kwargs)
+
+    def update(self, request, *args, **kwargs) -> Response:
+        """
+        update model instance if valid
+        """
+        self.validate()
+        return super().update(request, *args, **kwargs)
```

Comment:
Please use DRF's validation mechanism [1] we do not need more custom code, in fact you should also convert the bz_api_key parameter to a a serializer field (that does not get written back to the model/db)

[1] https://www.django-rest-framework.org/api-guide/validators/#validators

Reply:
Will have a look. Thank you for referencing!